File: python/src/code_smells/bloaters/primitive_obsession.py
class Order:
    def __init__(
        self,
        customer_name: str,
        customer_email: str,
        address: str,
        total_amount: float,
        currency: str,
    ):
        self.customer_name = customer_name
        self.customer_email = customer_email
        self.address = address
        self.total_amount = total_amount
        self.currency = currency

    def send_invoice(self):
        if '@' not in self.customer_email:
            raise ValueError('Email inválido')
        if not self.address:
            raise ValueError('No se ha indicado dirección')
        if self.total_amount <= 0:
            raise ValueError('El monto debe ser mayor que cero')
        print(f"Factura enviada a {self.customer_name} in {self.address} por {self.total_amount} {self.currency}")

File: go/code_smells/bloaters/primitive_obsession.go
package bloaters

import "fmt"

// Order demonstrates primitive obsession code smell
// Using primitive types instead of domain objects
type Order struct {
	CustomerName  string
	CustomerEmail string
	Address       string
	TotalAmount   float64
	Currency      string
}

// NewOrder creates a new order
func NewOrder(customerName, customerEmail, address string, totalAmount float64, currency string) *Order {
	return &Order{
		CustomerName:  customerName,
		CustomerEmail: customerEmail,
		Address:       address,
		TotalAmount:   totalAmount,
		Currency:      currency,
	}
}

// SendInvoice sends an invoice to the customer
func (o *Order) SendInvoice() error {
	// Validation logic scattered everywhere instead of being in domain objects
	if !containsAt(o.CustomerEmail) {
		return fmt.Errorf("email inválido")
	}
	if o.Address == "" {
		return fmt.Errorf("no se ha indicado dirección")
	}
	if o.TotalAmount <= 0 {
		return fmt.Errorf("el monto debe ser mayor que cero")
	}
	fmt.Printf("Factura enviada a %s in %s por %.2f %s\n",
		o.CustomerName, o.Address, o.TotalAmount, o.Currency)
	return nil
}

func containsAt(email string) bool {
	for _, c := range email {
		if c == '@' {
			return true
		}
	}
	return false
}

// Exercise: Refactor this code to use domain objects instead of primitives
// 1. Create an Email value object with validation
// 2. Create an Address value object
// 3. Create a Money value object to handle amount and currency together
// 4. Modify the Order to use these value objects
